pkg/provision/ec2.go

```diff
@@ -0,0 +1,203 @@
+package provision
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/google/uuid"
+)
+
+// EC2Provisioner contains the EC2 client
+type EC2Provisioner struct {
+	ec2Provisioner *ec2.EC2
+}
+
+// NewEC2Provioner creates an EC2Provisioner and initialises an EC2 client
+func NewEC2Provisioner(region, accessKey, secretKey string) (*EC2Provisioner, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String(region),
+		Credentials: credentials.NewStaticCredentials(accessKey, secretKey, ""),
+	})
+	svc := ec2.New(sess)
+	return &EC2Provisioner{ec2Provisioner: svc}, err
+}
+
+// Provision deploys an exit node into AWS
+func (p *EC2Provisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+	image, err := p.lookupAMI(host.OS)
+	if err != nil {
+		return nil, err
+	}
+
+	groupID, name, err := p.securityGroup(host.Additional["inlets-port"])
+	if err != nil {
+		return nil, err
+	}
+
+	runResult, err := p.ec2Provisioner.RunInstances(&ec2.RunInstancesInput{
+		ImageId:      image,
+		InstanceType: aws.String(host.Plan),
+		MinCount:     aws.Int64(1),
+		MaxCount:     aws.Int64(1),
+		UserData:     &host.UserData,
+		NetworkInterfaces: []*ec2.InstanceNetworkInterfaceSpecification{
+			{
+				DeviceIndex:              aws.Int64(int64(0)),
+				AssociatePublicIpAddress: aws.Bool(true),
+				DeleteOnTermination:      aws.Bool(true),
+				Groups:                   []*string{groupID},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.ec2Provisioner.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{runResult.Instances[0].InstanceId},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = p.ec2Provisioner.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{runResult.Instances[0].InstanceId},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: aws.String(*name),
+			},
+		},
+	})
+
+	return &ProvisionedHost{
+		ID:     *runResult.Instances[0].InstanceId,
+		Status: "",
+	}, nil
+}
+
+// Status returns the ID, Status and IP of the exit node
+func (p *EC2Provisioner) Status(id string) (*ProvisionedHost, error) {
+	var status string
+	s, err := p.ec2Provisioner.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(s.InstanceStatuses) == 0 {
+		return nil, fmt.Errorf("cannot retrieve status for instance: %s", id)
+	}
+
+	if *s.InstanceStatuses[0].InstanceStatus.Status == "ok" {
+		status = ActiveStatus
+	}
+
+	d, err := p.ec2Provisioner.DescribeInstances(&ec2.DescribeInstancesInput{
```

Comment:
Is DescribeInstanceStatus and DescribeInstances required here, if we know the instance will not be up immediately after the provisioning call?

This feels like it should belong in the Status function

Reply:
I am not sure what you mean, this is in the Status function.

Are you suggesting removing these checks and trying to get the instance and if it errors ignore the error and return a not ready status?

Reply:
My mistake, the diff must have been showing a merged view.